KAFKA-8774: Regex can be found anywhere in config value

wicknicks opened the pull request:
Signed-off-by: Arjun Satish <arjun@confluent.io>

The original method incorrectly used matcher.matches() instead of matcher.find(). The former method expects the entire string to match the regex, whereas the second one can find a pattern anywhere within the input string (which fits this use case more correctly).

Added unit tests.

Should be backported to 2.0.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

wicknicks committed:
KAFKA-8774: Regex can be found anywhere in config value

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@rhauch @avocader would appreciate a review. thanks!

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java:
```diff
@@ -474,6 +476,25 @@ public void testReverseTransformConfigs() {
+        testConfigProviderRegex("${file:/tmp/somefile.txt:somevar}");
+        testConfigProviderRegex("\"${file:/tmp/somefile.txt:somevar}\"");
+        testConfigProviderRegex("plain.PlainLoginModule required username=\"${file:/tmp/somefile.txt:somevar}\"");
+        testConfigProviderRegex("plain.PlainLoginModule required username=${file:/tmp/somefile.txt:somevar}");
```
How about adding some tests where multiple variables are used in a single value? I think the logic cares only if one can be found, but it still would be nice to verify.

Also, currently all of the variables are at the end -- how about some where they are at the beginning and middle with additional content besides the variable?

Also, what about other variable formats? Variables are of the form `${file:path:key}` (per [KIP-297](https://cwiki.apache.org/confluence/display/KAFKA/KIP-297%3A+Externalizing+Secrets+for+Connect+Configurations)), so maybe some without a path and very simple file, path, and key values.

wicknicks replied:
Added multiple cases here. Hope this helps. Thanks!

rhauch replied:
Excellent!

rhauch reviewed (commented):
Great find, @wicknicks! I like the minimalist PR, but have a few suggestions for being a bit more thorough with the new unit tests.

wicknicks committed:
KAFKA-8774: Add more cases for unit tests

Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch reviewed (approved):
LGTM, pending a green build. Thanks, @wicknicks!

wicknicks committed:
KAFKA-8774: Update system tests to cover case

Update system tests to cover case where config value contains
additional characters besides secret that requires regex pattern
to be found anywhere in the string (as opposed to complete match).

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
Updated a system test, here is a clean run https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2884/

rhauch reviewed (approved):
Thanks for updating the system test. LGTM!
